### PR TITLE
Parse error details during onboarding and revoking processes

### LINKED ISSUE #93: No access to StatusCode and ErrorMessage properties in RevokeException
I'd like to analyze the `StatusCode` property of `RevokeException` thrown after failed revoking. But both `StatusCode` and `ErrorMessage` properties are private: https://github.com/DKE-Data/agrirouter-sdk-dotnet-standard/blob/e62a86901f5760d0b3e2d0d71bf4e72b99975c51/agrirouter-sdk-dotnet-standard-api/Exception/RevokeException.cs#L21-L23

Is there any reason not to make them public?

Also, I noticed, that corresponding properties of `OnboardException` are public.

### LINKED ISSUE #95: Onboarding errors are not handled properly
In case of onboarding error, the onboarding API endpoint `api/v1.0/registration/onboard/request` returns a JSON structure with error description, such as (an example from documentation):
```json
{
  "error": {
    "code": "0110",
    "message": "Signing header is invalid, the request has timedout, or UTCTimestamp is not provided",
    "target": "",
    "details": []
  }
}
```
But `SecuredOnboardingService` doesn't parse this JSON in case of error, it simply puts the whole response body as a string to the `ErrorMessage` attribute of `OnboardException`: https://github.com/DKE-Data/agrirouter-sdk-dotnet-standard/blob/2cb8f903a49cfb5550544e01dc1b4b012c63334c/agrirouter-sdk-dotnet-standard-impl/Service/Onboard/SecuredOnboardingService.cs#L117-L119

It would be nice to parse the error and put response attributes to corresponding fields of `OnboardException`. 

Also, note that all the rest error attributes (`target`, `details`, and especially `code`) might be interesting to SDK users...

### PR DESCRIPTION
**Pull request by aparshin**
Fixes #95, fixes #93. This PR parses server error response during onboarding and revoking processes and puts all the parsed information inside the thrown exception.

Notes:
  * this is a breaking change. But I can make it backward compatible with the current version if it's critical to maintain backward compatibility.
  * I'm not sure about the naming of classes `OnboardError` and `OnboardException` (they are created during both onboarding and revoking). If the server returns errors in the same format (`{error: {code, message, target, details}}`) in all the endpoints, one can rename these classes and reuse them in other places.

### CODE REVIEW

File: agrirouter-sdk-dotnet-standard-impl/Service/Onboard/RevokeService.cs [hunk 1]
```diff
@@ -37,7 +37,7 @@ public RevokeService(Environment environment, HttpClient httpClient)
         /// </summary>
         /// <param name="revokeParameters">The parameters for the revoke process.</param>
         /// <param name="privateKey">The private key.</param>
-        /// <exception cref="RevokeException">Will be thrown if the revoking was not successful.</exception>
+        /// <exception cref="OnboardException">Will be thrown if the revoking was not successful.</exception>
```
Comment (saschadoemer) [inline agrirouter-sdk-dotnet-standard-impl/Service/Onboard/RevokeService.cs:40]: I do not like the concept of changing the specialized business exception to a more generic `OnboardException`. I would suggest having a common base class and keep the specialized exceptions.

**Review by saschadoemer — CHANGES_REQUESTED** [review]

File: agrirouter-sdk-dotnet-standard-api/Dto/Onboard/OnboardError.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+{
+    /// <summary>
+    ///     Data transfer object for the onboard and revoke error response.
+    /// </summary>
+    public class OnboardErrorResponse {
+        /// <summary>
+        ///     The error.
+        /// </summary>
+        [JsonProperty(PropertyName = "error")]
+        public OnboardError onboardError { get; set; }
```
[the remark this answers is not in the document]
Comment (aparshin) [reply agrirouter-sdk-dotnet-standard-api/Dto/Onboard/OnboardError.cs]: @saschadoemer My bad, fixed!

File: agrirouter-sdk-dotnet-standard-impl/Service/Onboard/RevokeService.cs (hunk 1 above)
Comment (aparshin) [reply agrirouter-sdk-dotnet-standard-impl/Service/Onboard/RevokeService.cs]: I created `OnboardRevokeExceptionBase` (any ideas about a better name?) base exception and inherited `OnboardException` and `RevokeException` from it.

**Comment by aparshin** [comment]
@saschadoemer Thank you for the review! I've fixed your notes, please check when you have a chance.

Additional note: I can't find in the documentation what exactly the AR server returns in case of error from `/iot/gateway/rest/measures/{{deviceAlternateId}}` and `/iot/gateway/rest/commands/{{deviceAlternateId}}` REST API endpoints, but I guess it might be the same JSON structure (cc @oliverrahner).

What do you think about parsing the error responses in `HttpMessagingService` and `FetchMessageService` similar to this PR?

**Comment by oliverrahner** [comment]
> * this is a breaking change. But I can make it backward compatible with the current version if it's critical to maintain backward compatibility.

I would really prefer being backward compatible here, even more so at it's not too much effort.

**Comment by aparshin** [comment]
@oliverrahner @saschadoemer
  * fixed conflicts with `main`
  * made it backward-compatible

File: agrirouter-sdk-dotnet-standard-api/Exception/CouldNotFetchMessagesException.cs [hunk 3]
```diff
@@ -14,14 +14,11 @@ public class CouldNotFetchMessagesException : System.Exception
...
+        public CouldNotFetchMessagesException(HttpStatusCode statusCode, string errorMessage): base(errorMessage)
         {
             StatusCode = statusCode;
-            ErrorMessage = errorMessage;
         }
 
-        private HttpStatusCode StatusCode { get; }
-
-        private string ErrorMessage { get; }
+        public HttpStatusCode StatusCode { get; }
```
Comment (saschadoemer) [inline agrirouter-sdk-dotnet-standard-api/Exception/CouldNotFetchMessagesException.cs:22]: I like that

File: agrirouter-sdk-dotnet-standard-api/Exception/CouldNotFetchMessagesException.cs [hunk 4]
```diff
@@ -14,14 +14,11 @@ public class CouldNotFetchMessagesException : System.Exception
         /// </summary>
         /// <param name="statusCode">-</param>
         /// <param name="errorMessage">-</param>
-        public CouldNotFetchMessagesException(HttpStatusCode statusCode, string errorMessage)
+        public CouldNotFetchMessagesException(HttpStatusCode statusCode, string errorMessage): base(errorMessage)
```
Comment (saschadoemer) [inline agrirouter-sdk-dotnet-standard-api/Exception/CouldNotFetchMessagesException.cs:17]:
```suggestion
        public CouldNotFetchMessagesException(HttpStatusCode statusCode, string message): base(message)
```

File: agrirouter-sdk-dotnet-standard-api/Exception/CouldNotSendMqttMessageException.cs [hunk 5]
```diff
@@ -19,6 +19,6 @@ public CouldNotSendMqttMessageException(MqttClientPublishReasonCode reasonCode,
             ReasonCode = reasonCode;
         }
 
-        private MqttClientPublishReasonCode ReasonCode { get; }
+        public MqttClientPublishReasonCode ReasonCode { get; }
```
Comment (saschadoemer) [inline agrirouter-sdk-dotnet-standard-api/Exception/CouldNotSendMqttMessageException.cs:22]: I like that 👍🏻

**Review by saschadoemer — CHANGES_REQUESTED** [review]
I have one last suggestion, everything else is fine.

File: agrirouter-sdk-dotnet-standard-api/Exception/CouldNotFetchMessagesException.cs [hunk 6]
```diff
@@ -14,14 +14,11 @@ public class CouldNotFetchMessagesException : System.Exception
         /// </summary>
         /// <param name="statusCode">-</param>
         /// <param name="errorMessage">-</param>
```
Comment (saschadoemer) [inline agrirouter-sdk-dotnet-standard-api/Exception/CouldNotFetchMessagesException.cs:16]:
```suggestion
        /// <param name="message">-</param>
```

**Review by saschadoemer — CHANGES_REQUESTED** [review]

**Review by saschadoemer — APPROVED** [review]

**Pull request merged**